synapse/storage/databases/main/schema/delta/59/12refresh_tokens.sql

```diff
@@ -0,0 +1,34 @@
+/* Copyright 2021 The Matrix.org Foundation C.I.C
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+-- Holds MSC2918 refresh tokens
+CREATE TABLE refresh_tokens (
+  id BIGINT PRIMARY KEY,
+  user_id TEXT NOT NULL,
+  device_id TEXT NOT NULL,
+  token TEXT NOT NULL,
+  -- When consumed, a new refresh token is generated, which is tracked by
+  -- this foreign key
+  next_token_id BIGINT REFERENCES refresh_tokens (id) ON DELETE CASCADE,
+  UNIQUE(token)
+);
+
+-- Add a reference to the refresh token generated alongside each access token
+ALTER TABLE "access_tokens"
+  ADD COLUMN refresh_token_id BIGINT REFERENCES refresh_tokens (id) ON DELETE CASCADE;
+
+-- Add a flag whether the token was already used or not
+ALTER TABLE "access_tokens"
+  ADD COLUMN used BOOLEAN NOT NULL DEFAULT FALSE;
```

Comment:
I'm a bit scared of adding a new column on this table with a default. Seems like Postgres 11 handles this properly: https://brandur.org/postgres-default but I have not tested.
The alternative is to make a null-able field instead.

Reply:
yeah, this is problematic in postgres 9.6 and 10. https://www.postgresql.org/docs/10/sql-altertable.html#SQL-ALTERTABLE-NOTES has some more info on it.

Best to make it nullable.

Reply:
I'd rather not have to worry about nullable boolean columns when querying if possible though.

Seems like we already did a lot of `ADD COLUMN ... NOT NULL DEFAULT ...` in the past:
```
synapse/storage/schema/main/delta/53/add_user_type_to_users.sql
19:ALTER TABLE users ADD COLUMN user_type TEXT DEFAULT NULL;

synapse/storage/schema/main/delta/27/forgotten_memberships.sql
26: ALTER TABLE room_memberships ADD COLUMN forgotten INTEGER DEFAULT 0;

synapse/storage/schema/main/delta/48/groups_joinable.sql
22:ALTER TABLE groups ADD COLUMN join_policy TEXT NOT NULL DEFAULT 'invite';

synapse/storage/schema/main/delta/59/14refresh_tokens.sql.postgres
17:ALTER TABLE "access_tokens"
18:  ADD COLUMN used BOOLEAN NOT NULL DEFAULT FALSE;

synapse/storage/schema/main/delta/59/11add_knock_members_to_stats.sql
16:ALTER TABLE room_stats_current ADD COLUMN knocked_members INT NOT NULL DEFAULT '0';
17:ALTER TABLE room_stats_historical ADD COLUMN knocked_members BIGINT NOT NULL DEFAULT '0';

synapse/storage/schema/main/delta/59/14refresh_tokens.sql.sqlite
17:ALTER TABLE "access_tokens"
18:  ADD COLUMN used BOOLEAN NOT NULL DEFAULT 0;

synapse/storage/schema/main/delta/56/hidden_devices.sql
18:ALTER TABLE devices ADD COLUMN hidden BOOLEAN DEFAULT FALSE;

synapse/storage/schema/main/delta/56/redaction_censor.sql
16:ALTER TABLE redactions ADD COLUMN have_censored BOOL NOT NULL DEFAULT false;

synapse/storage/schema/main/delta/58/08_media_safe_from_quarantine.sql.postgres
18:ALTER TABLE local_media_repository ADD COLUMN safe_from_quarantine BOOLEAN NOT NULL DEFAULT FALSE;

synapse/storage/schema/common/delta/58/00background_update_ordering.sql
19:ALTER TABLE background_updates ADD COLUMN ordering INT NOT NULL DEFAULT 0;

synapse/storage/schema/main/delta/58/08_media_safe_from_quarantine.sql.sqlite
18:ALTER TABLE local_media_repository ADD COLUMN safe_from_quarantine BOOLEAN NOT NULL DEFAULT 0;

synapse/storage/schema/main/delta/58/10federation_pos_instance_name.sql
20:ALTER TABLE federation_stream_position ADD COLUMN instance_name TEXT NOT NULL DEFAULT 'master';
```


Looking at the existing columns with defaults, seems like we don't have that many nullable columns with defaults:
```sql
select table_name, column_name, data_type, column_default, is_nullable
from information_schema.columns
where table_schema = 'public' 
  and column_default is not null
order by 1, 2;
```

|table_name                        |column_name         |data_type|column_default                                   |is_nullable|
|----------------------------------|--------------------|---------|-------------------------------------------------|-----------|
|access_tokens                     |used                |boolean  |false                                            |NO         |
|appservice_stream_position        |lock                |character|'X'::bpchar                                      |NO         |
|background_updates                |ordering            |integer  |0                                                |NO         |
|devices                           |hidden              |boolean  |false                                            |YES        |
|e2e_fallback_keys_json            |used                |boolean  |false                                            |NO         |
|e2e_room_keys_versions            |deleted             |smallint |0                                                |NO         |
|event_push_summary_stream_ordering|lock                |character|'X'::bpchar                                      |NO         |
|events                            |depth               |bigint   |0                                                |NO         |
|federation_stream_position        |instance_name       |text     |'master'::text                                   |NO         |
|groups                            |join_policy         |text     |'invite'::text                                   |NO         |
|instance_map                      |instance_id         |integer  |nextval('instance_map_instance_id_seq'::regclass)|NO         |
|local_media_repository            |safe_from_quarantine|boolean  |false                                            |NO         |
|push_rules                        |priority            |integer  |0                                                |NO         |
|received_transactions             |has_been_referenced |smallint |0                                                |YES        |
|redactions                        |have_censored       |boolean  |false                                            |NO         |
|room_memberships                  |forgotten           |integer  |0                                                |YES        |
|schema_version                    |lock                |character|'X'::bpchar                                      |NO         |
|stats_incremental_position        |lock                |character|'X'::bpchar                                      |NO         |
|user_directory_stream_pos         |lock                |character|'X'::bpchar                                      |NO         |
|users                             |admin               |smallint |0                                                |NO         |
|users                             |deactivated         |smallint |0                                                |NO         |
|users                             |is_guest            |smallint |0                                                |NO         |

Reply:
> Seems like we already did a lot of `ADD COLUMN ... NOT NULL DEFAULT ...` in the past:

Many of those are small tables, so can be rewritten during Synapse restart without major problems. I'm surprised to see a couple of quite large tables in that list, and I wonder if they caused us trouble at the time, but I don't think that changes my position on it here: I'd like us to avoid rewriting the access_tokens table if possible.

> Looking at the existing columns with defaults, seems like we don't have that many nullable columns with defaults:

To be clear, to avoid having to rewrite the table, you need to make it nullable with no default.